### ISSUE TITLE
Support for disabling error logging

### ISSUE THREAD

**Issue by zbarbuto** [marmelab/react-admin#8738]
**Is your feature request related to a problem? Please describe.**

The data provider will currently [log errors with console.error](https://github.com/marmelab/react-admin/blob/master/packages/ra-core/src/dataProvider/useDataProvider.ts#L114-L116) if `NODE_ENV !== 'production'`.

If you have a handful of tests for your error states which use e.g. 

```ts
useDataProvider(testDataProvider({
  getList: jest.fn().mockRejectedValue(new Error('Test Error'))
}))
```

The console can quickly get very spam-y as every single (intentional) error gets logged.

**Describe the solution you'd like**

I have a few suggestions (any of which I would be happy to submit a PR for)

1. (Simpler solution) Have another `process.env` flag that allows logs to be disabled (e.g. `process.env.NODE_ENV !== 'production' && process.env.NODE_ENV !== 'test'` or similar)
2. Add an option on either `useDataProvider` or the `DataProvider` itself to override logging (e.g. `useDataProvider({ logging: false })` or `<AdminContext dataProvider={{ getList: jest.fn(), logging: false }}>
3. Add support for a `logger` at some level (perhaps on `AdminContext`) which defaults to `console` for backwards compatibility and have the `useDataProvider` do a call to `useLogger()` or similar

**Describe alternatives you've considered**

Running my tests with `NODE_ENV='production'` but that seems like bad practice.

**Comment by slax57** [comment]
Hi, and thank you for this report.
You are raising an interesting point IMO.

1. First solution: although it seems like a good idea at first, I'm a bit worried it might make the debugging of tests more difficult for some configurations, where you have no way to also test the test-case in `development` mode
2. Second solution: I don't think adding an option on `useDataProvider` would work, because this hook is itself used in so many places in RA it would be hard to propagate the option from a single config entrypoint ; adding it on the `DataProvider` on the other hand might be a possible solution, as it would be retro-compatible, albeit admittedly a bit arguable to put a test-specific config here
3. Seems too complicated imo for such a specific case

Also, here is another alternative that we are using in the RA tests: for test that you expect to throw an error, you can mock the `console.error` implementation like so:
```ts
const mock = jest.spyOn(console, 'error').mockImplementation(() => {});
// ...
mock.mockRestore();
```
Or alternatively
```ts
const mock = jest.spyOn(console, 'error').mockImplementationOnce(() => {});
```

I'd like to have @fzaninotto  or @djhi 's opinion on this too. What do you guys think?

**Comment by zbarbuto** [comment]
Yes stubbing console.error is what I’m doing now but I don’t like it because it’s absorbing legitimate error logs from jest or react admin (such as the error when your mock data provider does not contain an array of data).

**Comment by fzaninotto** [comment]
Is there any case where we want to see these errors in tests? If the answer is no, one simple solution would be to log the dataProvider errors only in development:

```diff
-if (process.env.NODE_ENV !== 'production') {
+if (process.env.NODE_ENV === 'development') {
    console.error(error);
}
```

The other alternative would be to use a log channel, and letting developers disable logging in certain channels with another environment variable.

**Comment by slax57** [comment]
@fzaninotto 

> Is there any case where we want to see these errors in tests?

While writing the test? If they are not expected, they could help you debug a test that is not calling the DP the way you expect because of an error in the way the test is written.

To me, most of the time, I can write test cases that won't throw any unexpected DP error. The few cases where I can't are because I am actually testing an error case, or because I want to be quick and not bother implement a complete DP for instance... In such cases I don't mind mocking `console.error` once I've finished writing my test.

**Comment by fzaninotto** [comment]
@slax57 if you're in a test the `useDataProvider` throws an error you can catch. It's more robust to expect this error rather than to stup the console, isn't it?

https://github.com/marmelab/react-admin/blob/c0fcbb652cd7b9be7b024ca923aa4bf37851fb46/packages/ra-core/src/dataProvider/useDataProvider.ts#L131-L138

**Comment by slax57** [comment]
@fzaninotto I was thinking about tests cases where you do not expect any error, precisely, and you still run into one because the DP is wrong.

You may be correct that, for most cases, the error might be thrown and uncaught until caught by the test executor, which will display the error in the console.

But for errors that arise during the rendering phase of a component (e.g. after running a query to fetch the data), I'm worried such errors will be caught by the ErrorBoundary, which will make the test fail (because of assertions), but without printing the error in the console.
I must admit I'm not 100% sure it works that way. But if it does, this is a case where I would actually want to see the error in the console.

All in all, changing this condition is probably not that big of a deal because the case I identified is very specific and there are other ways to debug the tests (the console is only one of them). So I won't campaign further to prevent this change. Just wanted to mention I found a case where I would want to keep the logs.

**Comment by fzaninotto** [comment]
Thanks for your feedback. So it seems we can move forward with enabling error logging in the data provider only in `development`. 

The PR makes 2 line. If anyone wants to become a react-admin contributor without too much effort, that's your ticket!

**Comment by vytautassvirskas** [comment]
> Thanks for your feedback. So it seems we can move forward with enabling error logging in the data provider only in `development`.
> 
> The PR to solve this makes 2 line. If anyone wants to become a react-admin contributor without too much effort, that's your ticket!

hey 👋 
[fzaninotto](https://github.com/fzaninotto) I want to ask about that, you suggested adding changes in dataProvider, but in linked PR-#9262, the changes were made in use permissions, does this really solve the post-author issue?

**Comment by fzaninotto** [comment]
@vytautassvirskas Thanks for reaching out. You're right, data provider errors are still logged in tests:

https://github.com/marmelab/react-admin/blob/452d083bcdd5dfd87ace998ddb769b4f4322b19a/packages/ra-core/src/dataProvider/useDataProvider.ts#L112

I'm reopening this issue. We'll welcome a PR to fix it!
